gramlib/grammar.ml

```diff
@@ -220,16 +218,25 @@ type ('p, 'k, 'r) rel_prod = (unit, 'p, 'k, 'r) rel_prod0
 type ('s, 'i, 'k, 'r) any_symbols =
 | AnyS : ('s, 'p) ty_symbols * ('i, 'p, 'k, 'r) rel_prod0 -> ('s, 'i, 'k, 'r) any_symbols
 
-(** FIXME *)
-let rec symbols : type s p k r. (s, p) ty_symbols -> (s, k, r) ty_rule -> (s, unit, k, r) any_symbols =
-  fun accu r -> match r with
-  | TStop -> AnyS (Obj.magic accu, Rel0)
+type ('s, 'k, 'r) ty_belast_rule =
+| Belast : ('s, 'k, 'a -> 'r) ty_rule * ('s, 'a) ty_symbol -> ('s, 'k, 'r) ty_belast_rule
+
+(* unfortunately, this is quadratic, but ty_rules aren't too long
+ * (99% of the time of length less or equal 10 and maximum is 22
+ * when compiling Coq and its standard library) *)
+let rec get_symbols : type s k r. (s, k, r) ty_rule -> (s, unit, k, r) any_symbols =
+  let rec belast_rule : type s a k r. (s, k, r) ty_rule -> (s, a) ty_symbol -> (s, a -> k, r) ty_belast_rule =
+    fun r s -> match r with
+    | TStop -> Belast (TStop, s)
+    | TNext (r, s') ->
+       let Belast (r, s') = belast_rule r s' in
+       Belast (TNext (r, s), s') in
+  function
+  | TStop -> AnyS (TNil, Rel0)
   | TNext (r, s) ->
-    let AnyS (r, pf) = symbols (TCns (s, accu)) r in
-    AnyS (Obj.magic r, RelS (Obj.magic pf))
-
-let get_symbols : type s k r. (s, k, r) ty_rule -> (s, unit, k, r) any_symbols =
-  fun r -> symbols TNil r
+     let Belast (r, s) = belast_rule r s in
+     let AnyS (r, pf) = get_symbols r in
+     AnyS (TCns (s, r), RelS pf)
 
 let insert_tree (type s p k a) ~warning entry_name (gsymbols : (s, p) ty_symbols) (pf : (p, k, a) rel_prod) (action : k) (tree : (s, a) ty_tree) =
```

Comment:
15 lines below
```ocaml
                 (if entry_name <> "" then "" else "in ["^entry_name^"%s], ") ^
```
sounds a bit strange to me (how could `entry_name` be anything else than the empty string in the else branch?).

Reply:
Why not? It's given the name of the entry being extended in `levels_of_rules`.

Reply:
My question is only about the if expression, it seems to me that `if s <> "" then "" else s` will always be `""`

Reply:
So I just replaced the `<>` with `=`.